### ISSUE TITLE
BUG: linalg.expm: slower by a factor of ~4 in `scipy>1.13`

### ISSUE THREAD

**Issue by domna** [scipy/scipy#21078]
### Describe your issue.

scipy's expm function is getting slow for me when I upgrade to scipy 1.13.
For my machine I get this runtimes for the attached code: I get ~360 ms for scipy < 1.13 and 1800 ms for scipy >=1.13.

I checked the code but there was no relevant change with regard to the 1.12 versions, so it must probably in one of the functions which are called.

### Reproducing Code Example

```python
import time

import numpy as np
from scipy.linalg import expm

mat = np.array(
    [
        [0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 2.5774387j],
        [0.0 - 0.0j, 0.0 - 0.0j, 0.0 + 4.3402231j, 0.0 - 0.0j],
        [0.0 - 0.0j, 0.0 + 5.60367362j, 0.0 - 0.0j, 0.0 - 0.0j],
        [0.0 - 9.43618706j, 0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 0.0j],
    ]
)

start_time = time.time()
for _ in range(10000):
    expm(mat)
print(f"{(time.time() - start_time) * 1000} ms")
```

### Error message

```shell
No error message
```

### SciPy/NumPy/Python version and system information

```shell
1.13.1 2.0.0 sys.version_info(major=3, minor=9, micro=13, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SANDYBRIDGE MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 13.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 13.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld64
    name: gcc
    version: 11.3.0
  pythran:
    include directory: ../../../../../../private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-hy4ryy83/overlay/lib/python3.9/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
Python Information:
  path: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-run-d7dcw0c1/cp39-macosx_x86_64/build/venv/bin/python
  version: '3.9'
```

**Comment by nickodell** [comment]
Bisecting suggests that this is caused by ff99a9b900785c71bd049ff72f7eeecd9c08d31c, introduced by PR https://github.com/scipy/scipy/pull/20259.

I bisected between 1.12.0 and 1.13.1 using this script:

```
import time

import numpy as np
from scipy.linalg import expm

mat = np.array(
    [
        [0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 2.5774387j],
        [0.0 - 0.0j, 0.0 - 0.0j, 0.0 + 4.3402231j, 0.0 - 0.0j],
        [0.0 - 0.0j, 0.0 + 5.60367362j, 0.0 - 0.0j, 0.0 - 0.0j],
        [0.0 - 9.43618706j, 0.0 - 0.0j, 0.0 - 0.0j, 0.0 - 0.0j],
    ]
)

start_time = time.time()
for _ in range(10000):
    expm(mat)
duration = (time.time() - start_time) * 1000
print(f"{duration:.2f} ms")
if duration > 1300:
    exit(1)
```

Bisection log:

<details>

```
(scipy-dev) nodell@scipy-dev:~/scipy$ git bisect log
git bisect start
# good: [4edfcaa3ce8a387450b6efce968572def71be089] REL: SciPy 1.12.0 release commit [wheel build]
git bisect good 4edfcaa3ce8a387450b6efce968572def71be089
# bad: [44e4ebaac992fde33f04638b99629d23973cb9b2] REL: SciPy 1.13.1 release commit [wheel build]
git bisect bad 44e4ebaac992fde33f04638b99629d23973cb9b2
# good: [20d131e91b2974d01252f144247465645f6f2c1e] DOC: 1.12.0 release notes (#19628)
git bisect good 20d131e91b2974d01252f144247465645f6f2c1e
# good: [13739fbeac078713d4d732e1223764eed7fe6b45] TST: vectorised violation call
git bisect good 13739fbeac078713d4d732e1223764eed7fe6b45
# good: [03c05b0f1bd225759a66f631d14612e8e59bb726] Merge pull request #20218 from rgommers/add-sqrtm-xfail
git bisect good 03c05b0f1bd225759a66f631d14612e8e59bb726
# bad: [2a8c4ca6edceba5daaac60bbd213a84e7c03c361] Merge pull request #20290 from tylerjereddy/treddy_version_bump_rc2
git bisect bad 2a8c4ca6edceba5daaac60bbd213a84e7c03c361
# good: [f7d98c7fefc3f209e68f7b3bde8f2418ee4dc73a] Updated _specfun.pyx and moved specfun into special sublibrary (#20245)
git bisect good f7d98c7fefc3f209e68f7b3bde8f2418ee4dc73a
# good: [d25c443750ba817b38039818e2a2aac2b3c6b9c3] Merge pull request #19861 from jokabrink/fix-gh19659
git bisect good d25c443750ba817b38039818e2a2aac2b3c6b9c3
# bad: [5c8b806ee7d47e317ebb452a287b781af60acd94] Merge pull request #19777 from ev-br/bspline_insert
git bisect bad 5c8b806ee7d47e317ebb452a287b781af60acd94
# bad: [b520fe40de6bdc7993c569fa9694bf46c83e6212] MAINT: stats.wilcoxon: document addition of  support (#20266)
git bisect bad b520fe40de6bdc7993c569fa9694bf46c83e6212
# bad: [5e986a24f0ab03fbb382ed67ef15895024334481] BUG:linalg:Remove the 2x2 branch in expm (#20261)
git bisect bad 5e986a24f0ab03fbb382ed67ef15895024334481
# bad: [ff99a9b900785c71bd049ff72f7eeecd9c08d31c] BUG: linalg: fix `expm` for large arrays (#20259)
git bisect bad ff99a9b900785c71bd049ff72f7eeecd9c08d31c
# good: [8e5b96af972a57038a9899ce91ec46baaacc6607] ENH: integrate.solve_ivp: allow event `terminal` attribute to specify maximum number of events (#14771)
git bisect good 8e5b96af972a57038a9899ce91ec46baaacc6607
# first bad commit: [ff99a9b900785c71bd049ff72f7eeecd9c08d31c] BUG: linalg: fix `expm` for large arrays (#20259)
```

</details>

Comparison ff99a9b900785c71bd049ff72f7eeecd9c08d31c of to previous commit:

```
(scipy-dev) nodell@scipy-dev:~/scipy$ git rev-parse HEAD
8e5b96af972a57038a9899ce91ec46baaacc6607
(scipy-dev) nodell@scipy-dev:~/scipy$ python3 dev.py -n python ../foo/bug-21078.py 
Skipping build
579.36 ms
(scipy-dev) nodell@scipy-dev:~/scipy$ git checkout ff99a9b900785c71bd049ff72f7eeecd9c08d31c
Previous HEAD position was 8e5b96af9 ENH: integrate.solve_ivp: allow event `terminal` attribute to specify maximum number of events (#14771)
HEAD is now at ff99a9b90 BUG: linalg: fix `expm` for large arrays (#20259)
(scipy-dev) nodell@scipy-dev:~/scipy$ git ^C
(scipy-dev) nodell@scipy-dev:~/scipy$ python3 dev.py python ../foo/bug-21078.py 
💻  ninja -C /home/nodell/scipy/build -j6
ninja: Entering directory `/home/nodell/scipy/build'
[8/8] Linking target scipy/linalg/_matfuncs_expm.cpython-311-x86_64-linux-gnu.so
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
2190.16 ms
```

**Comment by lucascolley** [comment]
cc @thalassemia @ev-br

**Comment by ilayn** [comment]
There is only the change from `gemm` to NumPy `matmul` calls. In small arrays that extra might be significant but as size increases should not matter too much as 4x.

**Comment by domna** [comment]
> But we can reverse it if need be.

I think it would be nice to get the old performance back. If this is not too big of an issue it would be great if it could be reverted. I can also prepare a PR for this if you like.

**Comment by ilayn** [comment]
@domna Can you please post your speed comparison? I can't replicate this slowdown and also some folks previously had strange segfaults so I'm wondering if this is an architectural issue rather than the code. Because there is nothing fundamentally different code wise

**Comment by domna** [comment]
Sure, these are my results:

## `scipy == 1.12.0`

```
Array size 10: 1.2271404266357422 ms
Array size 100: 8.735895156860352 ms
Array size 1000: 59.48376655578613 ms
Array size 10000: 424.60083961486816 ms
Array size 20000: 677.1740913391113 ms
Array size 50000: 1766.974925994873 ms
```

## `scipy == 1.14.1`

```
Array size 10: 16.149044036865234 ms
Array size 100: 12.908935546875 ms
Array size 1000: 125.93698501586914 ms
Array size 10000: 1259.8068714141846 ms
Array size 20000: 2512.7041339874268 ms
Array size 50000: 6320.268154144287 ms
```

I can also test on different arch's if this helps (but at least I've seen it on three different ones so far, including arm/macOS and x86_64/linux)

**Comment by domna** [comment]
Sure, here is the comparion:

## `scipy == 1.12.0`
```
Matrix shape (2, 2): 607.6228618621826 ms
Matrix shape (4, 4): 263.9179229736328 ms
Matrix shape (8, 8): 465.89112281799316 ms
Matrix shape (16, 16): 812.2899532318115 ms
Matrix shape (32, 32): 2942.294120788574 ms
```

## `scipy == 1.14.1`

```
Matrix shape (2, 2): 878.425121307373 ms
Matrix shape (4, 4): 1133.5039138793945 ms
Matrix shape (8, 8): 1375.6070137023926 ms
Matrix shape (16, 16): 1666.9161319732666 ms
Matrix shape (32, 32): 3690.781831741333 ms
```


Matrices are generated with

```python
def gen_matrix(size):
    return np.random.rand(size, size) + 1j * np.random.rand(size, size)
```

I can also run this on different arch's if this helps (but the problem occurred on at least three different machines, including arm/macOS and x86_64/linux arch's)

**Comment by ilayn** [comment]
Thank you, I'll rewrite this in C just to see if it will fly faster. Because I can't still see what is slowing down the code with those changes.

**Comment by nickodell** [comment]
I did a benchmark of this function using my bisection script. Here are the timings for doing 10000 4x4 expm operations.

| Small matrix benchmark | Average (with 95% CI) |
|------------------------|-----------------|
| Pre regression (8e5b96af972a57038a9899ce91ec46baaacc6607)          | 580.4   ± 20.8 ms   |
| After regression (ff99a9b900785c71bd049ff72f7eeecd9c08d31c)       | 2254.2  ± 106.7 ms  |
| After fix (424708ed018cae6b6584d7d992940fd39f2ebcc0)              | 309.8   ± 21.0 ms  |

I find that this is 7x faster than the code with the performance regression, and almost 2x faster than the code before this performance regression.

Excellent work @eendebakpt and @ilayn!

**Comment by domna** [comment]
Awesome! Thank you for the great work @eendebakpt and @ilayn.
